### DISCUSSION TITLE
Same command for different chat type (private and group)

### DISCUSSION THREAD

**Discussion by alvinlys** [nksmnf/nestjs-telegraf#1218] [Q&A]
Hi, any chance using decorator/filters/guards so we can have the same command, e.g.: `Start()` but reply different in chat type of private and group?

**Comment by beautyfree** [comment]
same question

**Comment by armpogart** [comment]
Not the best idiomatic way in terms of NestJS, but this works:

```ts
@Injectable()
export class GroupChatGuard implements CanActivate {
    async canActivate(context: ExecutionContext): Promise<boolean> {
        const tgExecutionContext = TelegrafExecutionContext.create(context);
        const ctx = tgExecutionContext.getContext<Context>();
        const next = tgExecutionContext.getNext<(ctx: Context) => Promise<void>>();

        if (chat.type === 'group' || chat.type === 'supergroup') {
            return true;
        }
        
        await next(ctx);
        return false;
    }
}
```

```ts
@Injectable()
export class PrivateChatGuard implements CanActivate {
    async canActivate(context: ExecutionContext): Promise<boolean> {
        const tgExecutionContext = TelegrafExecutionContext.create(context);
        const ctx = tgExecutionContext.getContext<Context>();
        const next = tgExecutionContext.getNext<(ctx: Context) => Promise<void>>();

        if (chat.type === 'private') {
            return true;
        }
        
        await next(ctx);
        return false;
    }
}
```

Then use the guards on different controllers (or I guess methods, but I have only tried with different controllers). You can also do the same with only 1 of these guards, but in that case the order you've registered your controllers as a provider will matter.

P.S. For your specific case, just doing the check in the command handler and calling different services/methods will also suffice, and I would rather do that. The approach with these guards is best when you have a number of different handlers for group and private chats and want to properly separate the logic into 2 controllers.

**Comment by backrebital-lgtm** [comment]
Yes—register the command once and branch on Telegram’s chat type (private, group, or supergroup):

```ts
import { Command, Ctx, Update } from 'nestjs-telegraf';
import { Context } from 'telegraf';

@Update()
export class BotUpdate {
  @Command('start')
  async start(@Ctx() ctx: Context) {
    switch (ctx.chat?.type) {
      case 'private':
        return ctx.reply('Private-chat response');
      case 'group':
      case 'supergroup':
        return ctx.reply('Group-chat response');
      default:
        return; // updates without a chat (for example, channel updates)
    }
  }
}
```

If the handlers are large, keep this method as the dispatcher and move the private/group work into separate services. `ctx.chat.type` is the reliable discriminator; you don’t need separate command registrations.
